### PR TITLE
Users and groups management APIs for Samba

### PR DESCRIPTION
**Pull request by DavidePrincipi**
## Adding

- add-user, alter-user, remove-user 
- add-group, alter-group, remove-group

User attributes:
- password, 
- full name (displayName),
- locked state (account disabled),
- group membership,

Group attributes:
- description,
- members

## Testing

Install the module from `users-api` branch

    api-cli run add-internal-provider --data '{"image":"ghcr.io/nethserver/samba:users-api","node":1}'
    
Configure the domain from the UI page as usual

## Postponed

- Add "password does not expire" user flag 
- Domain password policy management

## See also

https://github.com/NethServer/ns8-core/pull/241 for the read side.

### CODE REVIEW

File: imageroot/actions/add-group/01validate_group [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python3
+
+#
+# Copyright (C) 2021 Nethesis S.r.l.
```
Comment (gsanchietti) [inline imageroot/actions/add-group/01validate_group:4]: Not important, but all files have 2021 as date :)

File: imageroot/actions/add-user/50add_user [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+groups = request['groups']
+password = request.get('password', '') # Empty string implies a random password is set
+full_name = request['full_name']
+locked = request.get('locked', False)
+
+sambatool_cmd = ['podman', 'exec', '-i', os.environ['MODULE_ID'], 'samba-tool']
+adduser_cmd = sambatool_cmd + ['user', 'create', user]
+
+if not password:
+    adduser_cmd += ['--random-password']
```
Comment (gsanchietti) [inline imageroot/actions/add-user/50add_user:40]: Is there a way to recover the random password?
I guess the admin just need to set it at later time.

File: imageroot/actions/alter-group/50alter_group [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import sys
+import subprocess
+import os
+
+request = json.load(sys.stdin)
+group = request['group']
+
+sambatool_cmd = ['podman', 'exec', os.environ['MODULE_ID'], 'samba-tool']
+
+if 'users' in request:
```
Comment (gsanchietti) [inline imageroot/actions/alter-group/50alter_group:34]: If `users` is not present, should the group emptied?

Comment (DavidePrincipi) [reply imageroot/actions/alter-group/50alter_group:34]: All attributes are optional (apart from `group`, the identifier). If an attribute is missing no change occurs. To empty a group set `users:[]`

File: imageroot/actions/add-user/50add_user (hunk 2 above)
Comment (DavidePrincipi) [reply imageroot/actions/add-user/50add_user:40]: Yes the admin must set it lately. I'm thinking about bulk users creation: Samba wants a complex password is set and provides `--random-password` to initialize the new user entry properly. UI is expected to pass a non-empty string to satisfy the JSON schema validator. The attribute is optional for bulk operations.

**Review by gsanchietti — APPROVED** [review]

**Pull request merged**